Завдання 11.7

rom-nov opened the pull request:
Було боляче але щось таки вийшло :)

rom-nov committed:
11-07 Add

rom-nov committed:
11-07 PR

rom-nov committed:
11-07 Додав mime_content_type()

rom-nov committed:
11-07 Почистив код

rom-nov committed:
11-07 Змінено Dockerfile. Додано ImageMagick.

rom-nov committed:
Merge branch 'main' into 11-07

roman-rtacademy commented on block-11/11-07/index.php:
```diff
@@ -0,0 +1,169 @@
+            <button class="form__btn" type="submit">Надіслати</button>
+        </div>
+        <?php
+            function safe_work( $message )
```
Типізація

roman-rtacademy commented on block-11/11-07/index.php:
```diff
@@ -0,0 +1,169 @@
+                }
+            }
+
+            function check_uploaded_file( $name, $max_size = 10485760 )
```
Типізація

roman-rtacademy commented on block-11/11-07/index.php:
```diff
@@ -0,0 +1,169 @@
+
+            function check_uploaded_file( $name, $max_size = 10485760 )
+            {
+                switch( true )
```
Цікаво, але краще зробити через набір `if( !someFunc() ) {...}`.

А ще краще - через chain методами класу

roman-rtacademy commented on block-11/11-07/index.php:
```diff
@@ -0,0 +1,169 @@
+
+            function is_empty( $file_name )
+            {
+                if( empty( $file_name ) )
```
Тобто це ж 
```
return empty( $file_name );
```
?

roman-rtacademy commented on block-11/11-07/index.php:
```diff
@@ -0,0 +1,169 @@
+                }
+            }
+
+            function is_empty( $file_name )
```
Типізація. Що повертає коли умова false?

roman-rtacademy commented on block-11/11-07/index.php:
```diff
@@ -0,0 +1,169 @@
+                exit;
+            }
+            check_uploaded_file( 'img' );
+            $img_origin = creat_img( 'img' );
```
А чому це все не в якойсь фукнції типу main() ?

roman-rtacademy reviewed (changes requested):
Див. коментарі

rom-nov committed:
11-07 правка зауважень PR

rom-nov commented:
Додав типізацію. Вирішив "пощюпати" об'єкти, спробував зробити ланцюг методів об'єкта.
Здається кроп-метод GDImages на великих файлах погано працює, чи то я щось не так роблю. Потім покажу.

rom-nov committed:
11-07 clear code

rom-nov committed:
11-07 clear code

rom-nov committed:
11-07 static class SaveFile

rom-nov committed:
11-07 __construct GDImageModifyFile

rom-nov committed:
11-07 __construct ControlLoadFile

rom-nov committed:
11-07 GDImageModifyFile -> destroy()

rom-nov committed:
11-07 __construct ControlLoadFile

rom-nov committed:
11-07
додав ImageModifyAbstract (виніс загальний функціонал)
змінив GDIImageModifyFile -> GDIImageModify (залишив функціонал GDImage)
видалив SaveFile

roman-rtacademy commented on block-11/11-07/ControlLoadFile.php:
```diff
@@ -0,0 +1,69 @@
+		$this -> is_empty() -> set_mime();
+	}
+
+	private function is_empty() : object
```
Щоб типізувати, можна писати заміть `: object` - `: self` або назву класу. 
Але якщо будеш робити extends, тоді краще імʼя інтерфейсу, який треба буде звісно також створити.

roman-rtacademy commented on block-11/11-07/index.php:
```diff
@@ -0,0 +1,58 @@
+        <?php
+        require 'ImageModifyAbstract.php';
+        require 'GDImageModify.php';
+        require 'ControlLoadFile.php';
```
Можеш спробувати автозавантаження через spl_autoload_register(), але це опціонально

roman-rtacademy commented on block-11/11-07/index.php:
```diff
@@ -0,0 +1,58 @@
+                    -> save_file( time() );
+
+				echo( '<img src="' . $img -> full_path() . '" width=auto height=auto>' );
+                $img -> destroy();
```
Так це можна перенести у 36 ряд, бо воно не видалить сам екземпляр класу для твого методу `full_path()`

roman-rtacademy commented on block-11/11-07/index.php:
```diff
@@ -0,0 +1,58 @@
+
+        //===== main script
+
+        if( empty( $_FILES ) & empty( $_POST ) )
```
`&` ?

rom-nov committed:
11-07
додав ControlLoadFileInterface, зробив типізацію у ControlLoadFile через інтерфейс.
у GDImageModify та ImageModifyAbstract замінив :object на :self у методах.
додав автозавантаження.

rom-nov committed:
11-07 strict_types in ControlLoadFileInterface

roman-rtacademy commented on block-11/11-07/index.php:
```diff
@@ -0,0 +1,56 @@
+        <?php
+        spl_autoload_register( fn( $class_name ) => require $class_name . '.php' );
+
+        function main( string $img_file ) : void
```
Опціонально. Спробуй розділяли логіку та view.

roman-rtacademy commented on block-11/11-07/ImageModifyAbstract.php:
```diff
@@ -0,0 +1,60 @@
+	{
+		if( $this -> get_width() < $size || $this -> get_height() < $size )
+		{
+			throw new Exception( 'Ширина та висота зображення має бути більше' . $size . 'px.' );
```
Після "більше" треба поставити пробіл

roman-rtacademy commented on block-11/11-07/ImageModifyAbstract.php:
```diff
@@ -0,0 +1,60 @@
+		return $this;
+	}
+
+	public function save_file( int|string $name, string $type = '.jpg', string $path = './', string $dir = 'data/' ) : self
```
А чому `int|string $name`? Може просто "string"?

roman-rtacademy commented on block-11/11-07/ImageModifyAbstract.php:
```diff
@@ -0,0 +1,60 @@
+<?php
```
declare( strict_types = 1);

rom-nov committed:
11-07 Main.php, ImageModifyInterface, розділив логіку та view

rom-nov committed:
11-07 переніс константи в інтерфейси, додав $image_path у Main

rom-nov committed:
11-07 переніс константи у Main та зробив передачу констант у методи через параметри

rom-nov committed:
11-07 переніс всі класи та інтерфейси у теку "lib" та додав namespace.

rom-nov commented:
Дуже сподобалась твоя реалізація, я у захваті від технічного класу Kernel, що містить всі налаштування в одному місті і передає їх у методи об'єктів, мега зручно. Вирішив теж переробити свій код, зробивши теж технічний клас і винести у нього логіку та налаштування. Бо спочатку я всі константи розкидав по інтерфейсах, що виявилось не дуже зручно, бо мусиш переходити по них щоб щось змінити, а з часом вже і не пригадаєш у якому інтерфейсі, які константи і налаштування. Додатково, теж все виніс у окрему теку "lib" + попрацював з namespace. Круто, я в захваті, php починає подобатися :)

rom-nov committed:
11-07 видалив у класі Main властивості file та img

rom-nov committed:
11-07 переніс реалізацію методу save_file ImageModifyAbstract -> GDImageModify
